### DISCUSSION TITLE
Confused about how to use speechbrain.utils.data_utils.batch_pad_right

### DISCUSSION THREAD

**Discussion by Gozzeh** [speechbrain/speechbrain#1656] [Q&A]
Apologies if this isn't the right place to ask this but the discourse says to ask code related questions here.

I'm trying to batch extract a set of embeddings because doing so one by one is painfully slow.  To this end I have written the following code

```
    sigsA= []

    for a in range( len( folder_analysis ) ):
        sigA, srA = torchaudio.load( folder_analysis[a] )
        sigsA.append( sigA )
        i = i + 1
        print( i )
    
    tensA = speechbrain.utils.data_utils.batch_pad_right( sigsA )
```
Every file is definitely mono and its a wave file and theyre all the same sample rate.

However when I call the batch_pad_right I get the error "OSError: Tensors should have same dimensions except for the first one".  So I'm clearly doing something wrong but I'm confused as to what.

Any help appreciated!

**Comment by Gastron** [comment]
The tensors need to have the time-dimension first. In your code, easiest would be to pass `channels_first=False` to `torchaudio.load`.

**Comment by Gozzeh** [comment]
Ok .. that makes sense.  And that then just leaves one more question and thats how do I batch extract the embeddings?

`embsA = classifier.encode_batch( tensA[0], tensA[1] )`

Calling this return "NotImplementedError: Only 2D, 3D, 4D, 5D padding with non-constant padding are supported for now".  So clearly still missing something.

**Comment by Gastron** [comment]
What model are you using? Can you post the traceback?

**Comment by Gozzeh** [comment]
Hi yes 

I'm using the pretrained ecapa-tdnn model as follows

`classifier = EncoderClassifier.from_hparams(source="speechbrain/spkrec-ecapa-voxceleb", savedir="pretrained_models/spkrec-ecapa-voxceleb")`

The full traceback is as follows:

```!continue
Traceback (most recent call last):

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\speechbrain\lobes\models\ECAPA_TDNN.py", line 480, in forward
    x = layer(x, lengths=lengths)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)

TypeError: forward() got an unexpected keyword argument 'lengths'


During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "c:\users\Gozzeh\speechbrain\speechbraintest.py", line 55, in <module>
    embsA = classifier.encode_batch( tensA[0], tensA[1] )

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\speechbrain\pretrained\interfaces.py", line 803, in encode_batch
    embeddings = self.mods.embedding_model(feats, wav_lens)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\speechbrain\lobes\models\ECAPA_TDNN.py", line 482, in forward
    x = layer(x)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\speechbrain\lobes\models\ECAPA_TDNN.py", line 76, in forward
    return self.norm(self.activation(self.conv(x)))

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\speechbrain\nnet\CNN.py", line 397, in forward
    x = self._manage_padding(

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\speechbrain\nnet\CNN.py", line 449, in _manage_padding
    x = F.pad(x, padding, mode=self.padding_mode)

  File "C:\Users\Gozzeh\anaconda3\lib\site-packages\torch\nn\functional.py", line 4207, in _pad
    raise NotImplementedError("Only 2D, 3D, 4D, 5D padding with non-constant padding are supported for now")

NotImplementedError: Only 2D, 3D, 4D, 5D padding with non-constant padding are supported for now```

**Comment by Gastron** [comment]
Huh, well following the traceback it's clear that in the Conv1D block it's trying to apply `torch.nn.functional.pad` with `padding_mode="reflect"`. But I have no idea why this problem comes up here and does not happen e.g. here: https://github.com/speechbrain/speechbrain/blob/9d56d50809d8745cc7dafa930aec554145be4028/recipes/VoxCeleb/SpeakerRec/speaker_verification_cosine.py#L43 

Maybe @hwidong-na or @nauman-daw has some idea?

**Comment by Gozzeh** [comment]
Just wondering if anyone else had any ideas on this?

Cheers!

**Comment by nauman-daw** [comment]
Hi @Gozzeh ,
I am guessing that it has to do something with the sampling rate. 
Are you re-sampling chunks on-the-fly? 
Are you using other sampling rate than 16k? I think somewhere in data preparation stage, sampling rate is kept 16k. I think there is also option to change it from yaml. If yes, please update those values with sampling rate of your data and check.

**Comment by Gozzeh** [comment]
hmm .. thats a good thought ... the source files may actually be 8kHz ...

**Comment by makaveli10** [comment]
Hello @Gozzeh , I am facing the same issue. were you able to get over this? If yes could you provide some information on that please.

**Comment by fabiocat93** [comment]
same issue here, has somebody of you found a solution maybe?
i can extract embeddings from one (mono) file, but not from a batch of (mono) files. has somebody found a solution to this?
